src/api/app/models/bs_request.rb

```diff
@@ -773,30 +767,34 @@ def change_review_state(new_review_state, opts = {})
         p[:description_extension] = superseded_by.to_s
         save!
         history.create(p)
-      elsif go_new_state # either no open reviews anymore or going back to review
-        if go_new_state == :new
-          history = HistoryElement::RequestAllReviewsApproved
-        elsif go_new_state == :declined
-          history = HistoryElement::RequestDeclined
-        elsif go_new_state != :review
-          raise "Unhandled state #{go_new_state} for history"
-        end
-        self.state = go_new_state if go_new_state
+        return
+      end
 
-        self.commenter = User.current.login
-        self.comment = opts[:comment]
-        self.comment = 'All reviewers accepted request' if go_new_state == :accepted
+      new_request_state = calculate_state_from_reviews
+      return if new_request_state == old_request_state
+      self.comment = opts[:comment]
+      self.state = new_request_state
+      self.commenter = User.current.login
+      if new_request_state == :new
+        history = HistoryElement::RequestAllReviewsApproved
+        self.comment = 'All reviewers accepted request'
+        save!
+        # pre-approved requests can be processed
+        BsRequestAutoAcceptJob.perform_later(id) if approver
+      elsif new_request_state == :review
+        save!
+      elsif new_request_state == :declined
+        history = HistoryElement::RequestDeclined
+        save!
```

Comment:
There is one subtle difference now which is that `change_review_state` saves the request. Before it didn't. Did you check that this doesn't have any side effect?

Reply:
see line 790 of the original - I just moved it into the if and duplicated it, so the code flow is more obvious